qiskit/pulse/filters.py

```diff
@@ -61,6 +64,60 @@ def filter_instructions(
     return filter_schedule
 
 
+@filter_instructions.register
+def _(
+    sched_blk: ScheduleBlock,
+    filters: List[Callable],
+    negate: bool = False,
+    recurse_subroutines: bool = True,
+) -> ScheduleBlock:
+    """A filtering function that takes a schedule_block and returns a schedule_block consisting of
+    filtered instructions.
+
+    Args:
+        sched_blk: A pulse schedule_block to be filtered.
+        filters: List of callback functions that take an instruction and return boolean.
+        negate: Set `True` to accept an instruction if a filter function returns `False`.
+            Otherwise the instruction is accepted when the filter function returns `False`.
+        recurse_subroutines: Set `True` to individually filter instructions inside of a subroutine
+            defined by the :py:class:`~qiskit.pulse.instructions.Call` instruction.
+
+    Returns:
+        Filtered pulse schedule_block.
+    """
+    from qiskit.pulse.transforms import inline_subroutines
+
+    target_sched_blk = sched_blk
+    if recurse_subroutines:
+        target_sched_blk = inline_subroutines(target_sched_blk)
+
+    def apply_filters_to_insts_in_scheblk(blk: ScheduleBlock) -> ScheduleBlock:
+        blk_new = ScheduleBlock.initialize_from(blk)
+        for element in blk.blocks:
+            if isinstance(element, ScheduleBlock):
+                inner_blk = apply_filters_to_insts_in_scheblk(element)
+                if len(inner_blk) > 0:
+                    blk_new.append(inner_blk)
+
+            elif isinstance(element, Instruction):
+                valid_inst = True
+                for filt in filters:
+                    valid_inst = np.logical_and(valid_inst, filt(element))
```

Comment:
Use of `np.logical_and` in this implementation doesn't make sense. In the schedule filter function, all instructions are first collected as a numpy array, then the higher order map function is used to apply a filter to all elements. This avoids use of python for-loop, which [might be slow](https://www.geeksforgeeks.org/python-map-vs-for-loop/).

In your implementation the logic evaluates each instruction in for-loop, and you can directly evaluate them by `all(filt(element) for filt in filters)`. The usage of for-loop looks okey for blocks, since this allows user to keep the nested structure of the input block and their alignment.

Reply:
Thank you! I updated as here 👉  https://github.com/Qiskit/qiskit-terra/pull/9971/commits/f222eab63d781244e74ab5c06ab238d1a9592706